pallene/checker.lua

```diff
@@ -164,7 +165,19 @@ end
 
 function Checker:add_module(name, is_main_mod)
     assert(type(name) == "string")
-    self.symbol_table:add_symbol(name, checker.Name.Module(name, is_main_mod))
+
+    -- To allow the name "string" to be used both as a module name and as a type name, module
+    -- symbols are aware of the type that they are shadowing.
+    local curr = self.symbol_table:find_symbol(name)
+
+    local shadowed_type
+    if curr and curr._tag == "checker.Name.Type" then
+        shadowed_type = curr
+    else
+        shadowed_type = false
+    end
```

Comment:
Would this also work?
```suggestion
    local shadowed_type = curr and curr._tag and curr.typ
```

If so, I think it's more direct, because you are only storing the necessary information.